File: solution/786. K-th Smallest Prime Fraction/786_02.cpp
class Solution {
public:
    vector<int> kthSmallestPrimeFraction(vector<int>& A, int k) {
        const int n = A.size();
        int a = 0, b = 0;
        
        double l = 0, r = 1, m;
        while (l < r) {
            m = l + (r - l) / 2;
            
            double fmax = 0;
            int cnt = 0;
            for (int i = 0, j = 0; i < n; ++i) {
                while (j < n && A[i] > m * A[j]) ++j;
                cnt += n - j;
                if (n == j) break;
                double f = A[i] * 1.0 / A[j];
                if (f > fmax) {
                    fmax = f;
                    a = i, b = j;
                }
            }
            
            if (cnt == k)
                return {A[a], A[b]};
            else if (cnt < k)
                l = m;
            else
                r = m;
        }
        return {};
    }
};

File: solution/786. K-th Smallest Prime Fraction/786_02.py
class Solution:
    def kthSmallestPrimeFraction(self, A: List[int], k: int) -> List[int]:
        n = len(A)
        a, b = 0, 0

        l, r = 0, 1
        while l < r:
            m = (l + r) / 2

            f_max = 0
            count = 0
            j = 0
            for i in range(n):
                while j < n and A[i] > m * A[j]:
                    j += 1
                if j == n:
                    break
                count += n - j
                f = A[i] / A[j]
                if f > f_max:
                    f_max = f
                    a, b = i, j

            if count == k:
                return [A[a], A[b]]
            elif count < k:
                l = m
            else:
                r = m
        return []
